### ISSUE TITLE
Warn about or automatically infer optimistic node ID

### ISSUE THREAD

**Issue by devknoll** [facebook/relay#119]
I've got a simple `getConfigs()` like:
```js
  getConfigs() {
    return [{
      type: 'FIELDS_CHANGE',
      fieldIDs: {
        step: this.props.step.id,
      },
    }, ];
  }
```

At first, I mistakenly tried using:
```js
  getOptimisticResponse() {
    return {
      step: {
        completed_at: this.props.complete ? null : moment.toISOString(),
      },
    };
  }
```

But nothing happened. After correctly adding the `id` field, everything was happy. It would be great if we could throw a warning in here (if there's a matching fieldID but the optimistic response doesn't contain an ID).

It might even be better if this could automatically be inferred. Have there been situations at FB where the ID in an optimistic response and the config have ever differed?

Happy to resolve this but want other thoughts first :smile:

**Comment by yuzhi** [comment]
I think that would be very useful here for anything with a `FIELDS_CHANGE` change config during an optimistic mutation. I don't remember run into any cases where the optimistic payload and the id in the config had a mismatch, so a infer could be useful for these optimistic payloads.

**Comment by devknoll** [comment]
@yuzhi Do you think it would be useful for the connection config types too? I haven't used them much yet, but they refer to a `parentID` which could theoretically also be missing from an optimistic mutation.

**Comment by yuzhi** [comment]
I don't think it would hurt. Overall, mutations could use a lot more warnings messages that can help guide developers with setting up their mutations. These changes doesn't have to be automagic, but should at least have useful messages, so someone who reads it will have an good idea of what to fix instead of just a silent failure. Mutation configs could also be easier to set up. There are definitely a lot of improvements that can be made to mutations.

**Comment by devknoll** [comment]
I think I agree regarding the automagic. A warning seems pretty good. I opened #144 for this, so I'll move any further discussion there!

**Comment by dminkovsky** [comment]
I just got bit by this problem. A warning would have saved me some time. When I figured out what was wrong, I realized that I had been had by this same issue some time ago too, doubling the pain :).

It was unintuitive for me that the optimistic payload should contain the ID given that the ID specified by the config. Does Relay even have the ability to handle IDs changing?

Furthermore, how do I "correctly" specific any optimistic payload for a mutation that has a `REQUIRED_CHILDREN` config that handles the creation of a new node, wherein the `REQUIRED_CHILDREN` fragment asks for the created ID? In such a situation, the ID comes with the response payload and cannot be predicted in the optimistic result.
